### PR TITLE
feat: Add shared network configuration types

### PR DESCRIPTION
**Pull request by noahbass**
PR draft:

Add common `Machine` and `Network` types in typescript for services to use as needed and to pass between services in a standard way.

TODO: add a serializer+deserializer. This can be accomplished by writing one ourselves or using an existing typescript library to do it automagically.

### CODE REVIEW

File: services/shared/network-configuration/network-configuration.ts [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+/**
+ * A configuration for a single machine in a computer network.
+ * Basically functions as a vertex in a graph.
+ */
+export class Machine {
+    readonly version: number = 1 // simple sequential API versioning
+    readonly machineId: string = uuidv4()
+    /**
+     * Anique software defined ip address
```
Comment (Samwisebuze) [inline services/shared/network-configuration/network-configuration.ts]: Spelling

File: services/shared/network-configuration/test/test.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
+import { assert } from 'chai'
+import { Machine, Network } from '../network-configuration'
+import { describe, it } from 'mocha'
+
+describe('Machine', () => {
```
Comment (Samwisebuze) [inline services/shared/network-configuration/test/test.ts]: Can we separate these into machine-test.js and netwrok-test.js? I have a feeling that their test will be changing quiet frequently and not always together.

**Review by Samwisebuze — CHANGES_REQUESTED** [review]
We should separate tests that won't **always** run together.

File: services/shared/network-configuration/test/test.ts (hunk 2 above)
Comment (noahbass) [reply services/shared/network-configuration/test/test.ts]: Yes good idea

**Comment by noahbass** [comment]
Libraries to consider for typescript serialization:
1. https://github.com/teq/serialazy
2. https://github.com/LabEG/Serializable

Also, I've come to the realization that having a `Set` of Machines is likely a bad api design - you would have to always have the `Machine` object to do any lookup/operation on a machine/network, where the only thing you really need to do a lookup/operation is the __uuid__ of the machine or network.

So, the design should probably be changed to...

```typescript
private machines: Map<string, Machine> // map of uuid to Machine
```

```typescript
private adjacentMachines: Map<string, Machine> // map of uuid to Machine
```

**Comment by noahbass** [comment]
7237dc3 Updated the data structure to store machines in a `Map`. This allows for more flexible and fast access for lookups and operations.

File: services/shared/network-configuration/network-configuration.ts [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+        const machineId = machine.machineId
+        this.machines.delete(machineId)
+
+        // Remove links between machine and other machines
+        for (const adjacent of adjacentMachines) {
+            // If the adjacent machine exists in the network,
+            // remove `machine` as being adjacent to it
+            const adjacentMachineId = adjacent.machineId
+            if (this.machines.has(adjacentMachineId)) {
+                adjacent.removeAdjacentMachine(machine)
```
Comment (Samwisebuze) [inline services/shared/network-configuration/network-configuration.ts]: Is this function safe? If it throws any errors it will break the cascade delete.

**Review by Samwisebuze — APPROVED** [review]

File: services/shared/network-configuration/network-configuration.ts (hunk 3 above)
Comment (noahbass) [reply services/shared/network-configuration/network-configuration.ts]: If `machine` isn't in `adjacent`'s adjacency list? `removeAdjacentMachine(...)` will never fail even if the `machine` doesn't exist. It uses Map.delete which can't fail. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/delete

**Comment by noahbass** [comment]
New commits in. Will rebase to reduce number of commits before we finally merge into staging.

**Review by Samwisebuze — APPROVED** [review]

**Pull request merged**